### PR TITLE
misc: expose handle print APIs

### PR DESCRIPTION
**Pull request by petkaantonov**
cc @bnoordhuis

### CODE REVIEW

**Comment by saghul** [comment]
This can be done easily using uv_walk. I'm -1.

File: docs/src/misc.rst [hunk 1]
```diff
@@ -251,3 +251,37 @@ API
...
+        stability guarantees.
+
+.. c:function:: void uv_print_active_handles(uv_loop_t* loop, FILE* stream)
+
+    This is the same as `uv_print_all_handles` except only active handles
+    are printed.
+
+    .. warning::
+        This function is meant for ad hoc debugging, there is no API/ABI
+        stability guarantees.
```
Comment (saghul) [inline docs/src/misc.rst]: We just can't do this.

**Comment by petkaantonov** [comment]
Yes but I would just copy paste libuv's code to io.js to do that then :|

**Comment by saghul** [comment]
> Yes but I would just copy paste libuv's code to io.js to do that then :|

Like the wise man once said: so? :-)

**Comment by petkaantonov** [comment]
also uv_walk doesn't iterate over internal handles, does it?

**Comment by saghul** [comment]
No, it doesn't. There is no reason to expose them, other than people shooting themselves in the foot.

I should have asked earlier, sorry: what do you need / want to accomplish? How does this help?

**Comment by petkaantonov** [comment]
For instance there is a bug in OS X where the internal async handle used for the select loop that backs streams to /dev segfaults in uv_async_event due to use-after-free - it took me quite long to realize it's not my async handle at all but the libuv internal async handle.

In general I see the workers implemenationg exposing many more bugs in libuv and this function would be very helpful - it is absolutely crucial that all handles are always closed when a worker terminates because we cannot rely on OS doing that for us anymore.

File: docs/src/misc.rst (hunk 1 above)
Comment (petkaantonov) [inline docs/src/misc.rst]: @bnoordhuis requested to add this to the documentation

**Comment by petkaantonov** [comment]
And the idea is that when someone is running io.js with the `--experimental-flags` enabled, they would not just get a random assertion failure when the loop is attempted to be closed with handles still on it - but they would instead get a message + list of *all* handles that were active at that point so that we don't need to ask someone to run gdb just so they can reproduce their issue again and call `uv_print_all_handles()` manually.

**Comment by saghul** [comment]
> For instance there is a bug in OS X where the internal async handle used for the select loop that backs streams to /dev segfaults in uv_async_event due to use-after-free - it took me quite long to realize it's not my async handle at all but the libuv internal async handle.
> 
> In general I see the workers implemenationg exposing many more bugs in libuv and this function would be very helpful - it is absolutely crucial that all handles are always closed when a worker terminates because we cannot rely on OS doing that for us anymore.

I can try to guess what those workers are, but I'm not familiar with your work there.

I assume you need to close a loop. libuv returns an error if you try to close a loop with active handles, so if you run into a problem is likely you're not closing the loop right. If you need to be absolutely certain that a loop is closed you can do the following:

* uv_walk all handles and call uv_close on them
* call uv_run with UV_RUN_DEFAULT mode
* call uv_loop_close, it must return 0, if it doesn't there is an internal problem in libuv.
* memory can now be safely released

**Comment by saghul** [comment]
> And the idea is that when someone is running io.js with the --experimental-workers flag enabled, they would not just get a random assertion failure when the loop is attempted to be closed with handles still on it - but they would instead get a message + list of all handles that were still associated with the loop at that point so that we don't need to ask someone to run gdb just so they can reproduce their issue again and call uv_print_all_handles() manually.

I understand that. But it makes me very uneasy of the precedent this sets. We promised backwards API compatibility and ABI stability, but now we'd add this with a note basically saying "we make no promises on these". Remember this is the stable libuv branch, so IMHO this is not acceptable.

@bnoordhuis, @piscisaureus thoughts?

**Comment by petkaantonov** [comment]
> I assume you need to close a loop. libuv returns an error if you try to close a loop with active handles, so if you run into a problem is likely you're not closing the loop right. If you need to be absolutely certain that a loop is closed you can do the following:

I actually had the "force close" logic using uv_walk already implemented but I removed it as it would still leak resources when someone needs custom callback to close additional resources. Instead general architecture for attaching handle closure logic when a thread exits is implemented. So when `uv_loop_close` is called and it returns non-zero, someone forgot to register a callback somewhere and it should be debugged.

> Remember this is the stable libuv branch, so IMHO this is not acceptable.

Yeah sorry I had no idea which is the correct branch to PR against.

**Comment by saghul** [comment]
> Yeah sorry I had no idea which is the correct branch to PR against.

I assume this is the correct branch since it will take time for libuv 2.0 to make an appearance, but in the meantime we have tried (and succeeded I might add! :-)) to not introduce undertainty into the codebase.

**Comment by saghul** [comment]
> I actually had the "force close" logic using uv_walk already implemented but I removed it as it would still leak resources when someone needs custom callback to close additional resources. Instead general architecture for attaching handle closure logic when a thread exits is implemented. So when uv_loop_close is called and it returns non-zero, someone forgot to register a callback somewhere and it should be debugged.

Hum. So you run all close callbacks, run the loop one last time and then call uv_loop_close? Note that closing the loop could take more than one iteration (at least that's true on Windows IIRC), so if you'd run the loop with UV_RUN_DEFAULT you'd prevent the loop from exiting.

**Comment by petkaantonov** [comment]
The general pattern is to call uv_close, and then wait until the callback is called before proceeding, e.g.

```cpp
  while (HandleCleanup* hc = handle_cleanup_queue_.PopFront()) {
    handle_cleanup_waiting_++;
    hc->cb_(this, hc->handle_, hc->arg_);
    delete hc;
  }

  while (handle_cleanup_waiting_ != 0)
    uv_run(event_loop(), UV_RUN_ONCE);
```

**Comment by petkaantonov** [comment]
@saghul  btw here is the irc conversation I had with bnoordhuis yesterday:

```
20:46 <petka__> btw is there a reason uv_print_all_handles is not exposed?
20:46 <bnoordhuis> it's just a debug tool i wrote to make my life easier
20:46  → dob_ joined  ⇐ nodweber and StoneCypher quit  
20:47 <petka__> yes it has been very handy, but annoying to add to uv export everytime I need it :D
20:47 <bnoordhuis> oh, you can always call it from inside gdb
20:47 <bnoordhuis> that's my main use case, really
20:48 <petka__> I was thinking of printing all handles to stderr when a worker terminates and its loop still has stuff on it
20:48 <petka__> and abort
20:48 <petka__> cos basically at that point the loop is supposed to be completely empty (not counting the 2 internal handles that are always there)
20:49 <bnoordhuis> i guess that's convenient. if you're going to file a pr, please add a comment in uv.h and the docs explaining that the function is not part of the normal api/abi stability guarantees
20:49 <petka__> cool
```

**Comment by bnoordhuis** [comment]
I personally don't see anything wrong with making exemptions for certain functions, provided they're clearly advertised as such (which they are.)

If it's controversial, maybe put them in a uv-experimental.h header?  I defer to @saghul, I don't really have a dog in this fight.

**Comment by petkaantonov** [comment]
Yeah I don't need it to be in a stable header so @saghul is the experimental header fine for you?

**Comment by saghul** [comment]
Let me think about this some more.

**Comment by petkaantonov** [comment]
Also, if it's in experimental header it could also be documented on some hard-to-find experimental api page rather than in mixed with stable api docs ;p

**Comment by equalsraf** [comment]
I've bumped into similar unclosed handle issues too. Printing this information is not very useful for handling errors at runtime, instead of adding an API for this what would help is some instrumentation for debugging it e.g. in debug builds log this somewhere when calling `uv_loop_close`.

@saghul already has a nice [description on how](http://stackoverflow.com/questions/25615340/closing-libuv-handles-correctly?answertab=votes#tab-top) to handle this type of errors, that would fit well into the documentation (or maybe its already there I missed it).

**Comment by saghul** [comment]
Sorry for the absurd delay here... I had a change of hear, lets land this :-) Can you please make sure the functions are not defined inside "#ifndef NDEBUG"? Thanks!

**Comment by petkaantonov** [comment]
done

**Comment by saghul** [comment]
Thanks! LGTM.

**Comment by saghul** [comment]
Thanks Petka, landed in 6490c50, and sorry for having this almost rot here :-S

(I added a ".. versionadded" to the docs part)

**Pull request closed without merging**